_drafts/kotlin-coroutines-101.md

```diff
@@ -0,0 +1,915 @@
+  coroutineScope {
+    launch(dispatcher) {
+      workingConsciousness()
+    }
+    launch(dispatcher) {
+      takeABreak()
+    }
+  }
+}
+```
+
+Now, the log shows that the `takeABreak` coroutine had the chance to execute, even if the `workingConsciousness` runs forever and we have a single thread:
+
+```text
+09:02:49.302 [main] INFO CoroutinesPlayground - Starting the morning routine
+09:02:49.376 [DefaultDispatcher-worker-1 @coroutine#1] INFO CoroutinesPlayground - Working
+09:02:49.382 [DefaultDispatcher-worker-1 @coroutine#2] INFO CoroutinesPlayground - Taking a break
+09:02:50.387 [DefaultDispatcher-worker-1 @coroutine#2] INFO CoroutinesPlayground - Break done
+-- Running forever --
+```
+
+We can obtain the same log also using the `workingHard` coroutine, adding a thread to the thread pool:
+
+```kotlin
+@OptIn(ExperimentalCoroutinesApi::class)
+suspend fun workingHardRoutine() {
+  val dispatcher: CoroutineDispatcher = Dispatchers.Default.limitedParallelism(2)
+  coroutineScope {
+    launch(dispatcher) {
+      workingHard()
+    }
+    launch(dispatcher) {
+      takeABreak()
+    }
+  }
+}
+```
+
+Since we have two threads and two coroutines, the concurrency degree is now two. As usual, the log confirms the theory: `coroutine#1` executes on `DefaultDispatcher-worker-1`, and `coroutine#2` executes on `DefaultDispatcher-worker-2`.
+
+```text
+13:40:59.864 [main] INFO CoroutinesPlayground - Starting the morning routine
+13:40:59.998 [DefaultDispatcher-worker-1 @coroutine#1] INFO CoroutinesPlayground - Working
+13:41:00.003 [DefaultDispatcher-worker-2 @coroutine#2] INFO CoroutinesPlayground - Taking a break
+13:41:01.010 [DefaultDispatcher-worker-2 @coroutine#2] INFO CoroutinesPlayground - Break done
+-- Running forever --
+```
+
+**Cooperative scheduling forces us to be very careful when designing our coroutines**. Suppose a coroutine performs an operation blocking the underlying thread, such as a JDBC call. In that case, it blocks the thread from executing any other coroutine.
+
+For this reason, the library allows us to use different dispatchers for different operations. The main ones are:
+
+1. `Dispatchers.Default` is the default dispatcher used by the library. It uses a thread pool with a number of threads equal to the number of available processors. It's the right choice for CPU-intensive operations.
+2. `Dispatchers.IO` is the dispatcher used for I/O operations. It uses a thread pool with a number of threads equal to available processors or, at most 64. It's the right choice for I/O operations, such as network calls or file operations.
+3. Dispatcher created from a thread pool: It's possible to make our instance of `CoroutineDispatcher` using a thread pool. We can easily use the `asCoroutineDispatcher` extension function of the `Executor` interface. However, be aware that it's our responsibility to close the underlying thread pool when we don't need it anymore:
+
```

Comment:
I think we can add that if we have both CPU-intensive parts and blocking parts, we should use both, and make sure to launch "normal" coroutines on the default dispatchers, and blocking code on the IO dispatcher.